web-api-db-ctrl/src/lib.rs

```diff
@@ -0,0 +1,524 @@
+            &mut tx,
+            &"public".parse()?,
+            &PUBLIC_SCHEMA_MIGRATOR,
+        )
+        .await?;
+
+        tx.commit().await?;
+
+        // We run this _before_ we release the lock but it will
+        // run on concurrently on  multiple different connections.
+        //
+        // For this we can have the same guarantees with multi tenant as we
+        // currently have with single tenant.
+        //FIXME: There is a limit to how well this scales.
+        info!("start tenant db schema migrations");
+        let failures = self.run_all_db_migrations(false).await?;
+
+        unlock_lock_id(&mut conn, MIGRATION_LOCK_ID).await?;
+
+        //TODO we need to decide how to handle partial failure
+        if !failures.is_empty() {
+            for (tenant_id, error) in failures {
+                error!({ %tenant_id, %error }, "migration failed");
+            }
+            bail!("all tenant migrations failed");
+        }
+
+        Ok(legacy_tenant_id)
+    }
+
+    async fn initialize_legacy(
+        &self,
+        tx: &mut Transaction<'_, Postgres>,
+    ) -> Result<TenantId, Error> {
+        let legacy_tenant_id = sqlx::query_as::<_, (TenantId,)>(
+            "SELECT tenant_id FROM tenant WHERE is_legacy_tenant FOR UPDATE;",
+        )
+        .fetch_optional(&mut *tx)
+        .await?;
+
+        let legacy_tenant_id = if let Some((legacy_tenant_id,)) = legacy_tenant_id {
+            legacy_tenant_id
+        } else {
+            let new_id = TenantId::random();
+            create_tenant_with_empty_schema(tx, new_id, true).await?;
+            info!({tenant_id = %new_id}, "created new legacy tenant");
+            new_id
+        };
+
+        if does_table_exist(&mut *tx, "public", "documents").await? {
+            let tenant = QuotedIdentifier::db_name_for_tenant_id(legacy_tenant_id);
+
+            if does_schema_exist(&mut *tx, tenant.as_unquoted_str()).await? {
+                bail!("database has both public legacy schemas and a migrated schema, this should be impossible");
+            }
+
+            info!("moving legacy tenant from public schema to {tenant}");
+
+            let query = format!(
+                "ALTER SCHEMA public RENAME TO {tenant};
+                -- create a new public schema, wo do not grant rights to PUBLIC
+                CREATE SCHEMA public;
+                -- revoke privileges from public
+                REVOKE ALL ON SCHEMA {tenant} FROM PUBLIC;
+                -- probably unneeded but make sure it's owned by the admin user
+                ALTER SCHEMA {tenant} OWNER TO CURRENT_USER;"
+            );
+            tx.execute_many(query.as_str())
+                .try_for_each(|_| future::ready(Ok(())))
+                .await?;
+        }
+        Ok(legacy_tenant_id)
+    }
+
+    /// Allows using the admin user as `web-api-mt` user.
+    #[instrument(skip(self), err)]
+    pub async fn admin_as_mt_user_hack(&self) -> Result<(), Error> {
```

Comment:
Just to avoid problems: This has nothing to do with the legacy tenant handling.

This exist because the code already assumes there are two users:

1. `admin` which owns the schemas/structure of the db and is e.g. used to create tenants and run schema migrations
2. `web-api-mt` which only is allowed to use schemas

I'm not sure if I like this hack, but managing permissions for this later one is a bit of a pain. And changing our deployment in the future to have two users from the get to go shouldn't be too hard :thinking: